
# 📊 Projeto de Segmentação de Clientes — E-commerce (RFM, PCA, Clustering, Business)

Neste projeto, utilizei o dataset **Ecommerce**, que contém informações como *InvoiceNo, StockCode, Country*, entre outras variáveis transacionais, com o objetivo de segmentar clientes e extrair insights de negócio.

O fluxo do projeto inclui:

- **Data Cleaning**: remoção de valores inválidos, tratamento de dados ausentes e ajustes necessários para análise.
- **Feature Engineering**: criação das métricas de **RFM (Recency, Frequency, Monetary)**.
- **Análise Exploratória (EDA)** do dataset original e das variáveis RFM.
- **Análise estatística** descritiva tanto do dataset quanto dos clusters gerados.
- Uso de **winsorization** para lidar com caudas longas nas visualizações.
- **Tratamento de outliers** utilizando percentis e o algoritmo **DBSCAN**.
- Aplicação de **PCA** para redução de dimensionalidade e visualização em 2D.
- Treinamento do modelo **K-Means** para segmentação de clientes.
- Visualizações dos clusters no espaço original e no espaço reduzido pelo PCA.
- Análise detalhada dos clusters obtidos, incluindo:
  - identificação de clientes que cancelaram,
  - criação de *naming* dos grupos (**VIP, Regulares e Inativos**),
  - propostas de estratégias de retenção e engajamento para cada segmento.

As principais bibliotecas utilizadas foram: **pandas**, **numpy**, **matplotlib**, **seaborn** e módulos do **scikit-learn**, incluindo **PCA**, **DBSCAN** e **KMeans**.

O objetivo final é demonstrar um pipeline completo de análise e segmentação de clientes, combinando estatística, machine learning e interpretação de negócio.

## 📁Colunas do dataset :



| Coluna        | Descrição |
|--------------|-----------|
| **InvoiceNo** | Identificador único da fatura (nota fiscal). Valores que começam com **"C"** indicam **cancelamentos ou devoluções**. |
| **StockCode** | Código único do produto no sistema da empresa. |
| **Description** | Descrição textual do produto vendido. |
| **Quantity** | Quantidade de unidades do produto na transação.  |
| **InvoiceDate** | Data e hora em que a transação foi realizada. |
| **UnitPrice** | Preço unitário do produto.|
| **CustomerID** | Identificador único do cliente. Pode conter valores ausentes para clientes não cadastrados. |
| **Country** | País onde o cliente está localizado. |

## 📌 Contexto e Propósito do Projeto

Em cenários de e-commerce, compreender o comportamento dos clientes é essencial para tomar decisões estratégicas baseadas em dados. Bases transacionais costumam conter padrões ocultos de consumo que, quando bem explorados, permitem segmentar clientes e direcionar ações de negócio de forma mais eficiente.

O propósito deste projeto é **identificar grupos de clientes com comportamentos semelhantes**, utilizando técnicas de análise exploratória, estatística e aprendizado não supervisionado. A partir dessa segmentação, torna-se possível:

- **Separar clientes em clusters distintos**, com base em padrões de recência, frequência e valor gasto (RFM);
- **Analisar o perfil de cada grupo**, entendendo nível de engajamento e valor gerado;
- **Extrair insights acionáveis** para apoiar decisões de marketing e retenção;
- **Criar estratégias específicas para cada cluster**, em vez de abordagens genéricas;
- **Melhorar a retenção de clientes ativos**, reduzindo churn;
- **Reativar clientes inativos**, por meio de campanhas direcionadas;
- **Estimular a migração de clientes regulares para o grupo VIP**, aumentando frequência e ticket médio;
- Apoiar estratégias para **atração de novos clientes**, usando o perfil dos melhores consumidores como referência.

Assim, o projeto demonstra como técnicas de **clustering e análise estatística** podem ser usadas para transformar dados brutos de e-commerce em conhecimento estratégico, com impacto direto em retenção, engajamento e geração de valor para o negócio.

In [20]:
import pandas as pd

In [21]:
df = pd.read_csv('../data/raw.csv', encoding='latin1')
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Quantity,493536.0,7.797514,47.760585,-9600.00,1.00,3.0,10.00,12540.00
UnitPrice,493536.0,4.553079,82.082705,-11062.06,1.25,2.1,4.13,17836.46
CustomerID,358456.0,15326.634856,1711.586710,12347.00,13911.00,15358.0,16811.00,18287.00


Podemos notar através da função describe que temos valores **negativos** em Quantity (Quantidade) e UnitPrice (Preço unitário), isso quer dizer que em Quantity há cancelamentos e em UnitPrice há erros já que preço não pode ser negativo.

Além disso, nota-se que os valores máximos de Quantity e UnitPrice estão **muito acima do percentil 75%** , evidenciando uma distribuição **altamente assimétrica**. Enquanto 75% dos registros apresentam Quantity ≤ 10 e UnitPrice ≤ 4,13, existem observações com valores extremamente elevados, que se distanciam significativamente da média.

Apesar da média estar próxima do percentil 75%, indicando uma **grande concentração de valores baixos**, a presença desses valores extremos exerce forte influência nas estatísticas descritivas, caracterizando a **existência de clientes de alto consumo**.

In [22]:
df.describe(include='object').T

,count,unique,top,freq
InvoiceNo,493536,23384,573585,1114
StockCode,493536,4064,85123A,2110
Description,492082,4188,WHITE HANGING HEART T-LIGHT HOLDER,2168
InvoiceDate,493536,21148,10/31/2011 14:41,1114
Country,493536,38,United Kingdom,457492


Através deste describe conseguimos descobrir que o **produto mais vendido foi o WHITE HANGING HEART T-LIGHT HOLDER**,  dentre os 38 países **United Kingdom (Reino unido) foi o país que mais foram feito vendas**, o cliente de **InvoiceNo 573585 foi o cliente que mais comprou de uma só vez levando 1114 itens.**

### Verificando cancelamentos e guardando no dataframe df_cancelamentos :

In [23]:
df_cancelamentos = df[df['InvoiceNo'].str.startswith('C')]
df_cancelamentos.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
121,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
122,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
123,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
124,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
125,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,12/1/2010 10:24,3.45,17548.0,United Kingdom


Podemos perceber agora que temos registros especiais em StockCode, como D em Discount, como iremos fazer segmentação e análise de retenção para clientes podemos excluir também estes registros especiais do dataset.

### Excluindo do dataframe os registros de cancelamento ou devolução :

In [24]:
df = df[~df['InvoiceNo'].str.startswith('C')]

### Verificando os registros especiais :

In [25]:
reg_especiais = df['StockCode'].str.match(r'^[A-Za-z]')

df[reg_especiais]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
36,536370,POST,POSTAGE,3,12/1/2010 8:45,18.00,12583.0,France
241,536403,POST,POSTAGE,1,12/1/2010 11:27,15.00,12791.0,Netherlands
952,536527,POST,POSTAGE,1,12/1/2010 13:04,18.00,12662.0,Germany
1620,536544,DOT,DOTCOM POSTAGE,1,12/1/2010 14:32,569.77,NaN,United Kingdom
1980,536569,M,Manual,1,12/1/2010 15:35,1.25,16274.0,United Kingdom
...,...,...,...,...,...,...,...,...
492889,581494,POST,POSTAGE,2,12/9/2011 10:13,18.00,12518.0,Germany
493213,581498,DOT,DOTCOM POSTAGE,1,12/9/2011 10:26,1714.17,NaN,United Kingdom
493401,581570,POST,POSTAGE,1,12/9/2011 11:59,18.00,12662.0,Germany
493418,581574,POST,POSTAGE,2,12/9/2011 12:09,18.00,12526.0,Germany


### retirando - os do dataframe :

In [26]:
df = df[df['StockCode'].str.match(r'^\d')]

### Verificando dados faltantes :

In [27]:
print(df.info())

print('\n Dados faltantes: \n\n',df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
Index: 483366 entries, 0 to 493535
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    483366 non-null  object 
 1   StockCode    483366 non-null  object 
 2   Description  481927 non-null  object 
 3   Quantity     483366 non-null  int64  
 4   InvoiceDate  483366 non-null  object 
 5   UnitPrice    483366 non-null  float64
 6   CustomerID   349526 non-null  float64
 7   Country      483366 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.2+ MB
None

 Dados faltantes: 

 InvoiceNo           0
StockCode           0
Description      1439
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133840
Country             0
dtype: int64


Foi verificado que possuimos **dados faltantes** em **Description (1439 registros)**  e em **CustomerID (133840)**. Nestes casos não podemos imputar nada, CustomerID é um valor aleatório e Description é um produto.
Verificando os tipos de dados temos que nos atentar a colocar o InvoiceDate para formato de data (datetime), irei fazer isso no código abaixo.

In [28]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

### Verificando os dados faltantes :

In [29]:
df_missing = df[df['CustomerID'].isnull() | df['Description'].isnull()]
df_missing

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
458,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1249,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1250,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1251,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1252,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
493208,581498,85049e,SCANDINAVIAN REDS RIBBONS,4,2011-12-09 10:26:00,3.29,NaN,United Kingdom
493209,581498,85099B,JUMBO BAG RED RETROSPOT,5,2011-12-09 10:26:00,4.13,NaN,United Kingdom
493210,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,2011-12-09 10:26:00,4.13,NaN,United Kingdom
493211,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,2011-12-09 10:26:00,4.96,NaN,United Kingdom


É possível perceber que, exceto no primeiro registro que só tem um InvoiceNo único, **todos os outros registros com customerID NAN vem de um mesmo InvoiceNo, ou seja, de uma nota fiscal**. Isso me fez perguntar se teve algum registro onde InvoiceNo combina e existe um CustomerID não nulo (Pode ter ocorrido um erro), se isso ocorre podemos preencher os dados faltantes .

### Excluindo valores nulos :

In [30]:
df = df.dropna(subset=['CustomerID', 'Description'])
df.shape

(349526, 8)

In [31]:
df.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

### Verificando se ha registros onde quatidade é negtivo :

In [32]:
df[df['Quantity'] <= 0 ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


**Não há registros com quantidade negativa**

### Verificando se há UnitPrice negativo

In [33]:
df[df['UnitPrice'] <= 0 ].shape

(21, 8)

**33 registros onde UnitPrice = 0, precisamos apagar** pois não é possível o preço ser negativo.

In [34]:
df.duplicated().sum()
df_duplicates = df[df.duplicated(keep=False)]
df_duplicates

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
321,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
325,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
330,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
353,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
357,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
...,...,...,...,...,...,...,...,...
493348,581538,22068,BLACK PIRATE TREASURE CHEST,1,2011-12-09 11:34:00,0.39,14446.0,United Kingdom
493362,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,2011-12-09 11:34:00,2.49,14446.0,United Kingdom
493365,581538,22992,REVOLVER WOODEN RULER,1,2011-12-09 11:34:00,1.95,14446.0,United Kingdom
493372,581538,22694,WICKER STAR,1,2011-12-09 11:34:00,2.10,14446.0,United Kingdom


Estes registros acima são registros onde há **duplicados**, ou seja, estão iguais a outros registros, e portanto devem ser **excluídos**.

In [35]:
df = df.drop_duplicates()
df.shape

(344771, 8)

## Feature engineering :

### Criando a coluna Revenue (Lucro) :

In [36]:
df['Revenue'] = df['Quantity'] * df['UnitPrice']
df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047.0,United Kingdom,54.08
1,536367,22745,POPPY'S PLAYHOUSE BEDROOM,6,2010-12-01 08:34:00,2.10,13047.0,United Kingdom,12.60
2,536367,22748,POPPY'S PLAYHOUSE KITCHEN,6,2010-12-01 08:34:00,2.10,13047.0,United Kingdom,12.60
3,536367,22749,FELTCRAFT PRINCESS CHARLOTTE DOLL,8,2010-12-01 08:34:00,3.75,13047.0,United Kingdom,30.00
4,536367,22310,IVORY KNITTED MUG COSY,6,2010-12-01 08:34:00,1.65,13047.0,United Kingdom,9.90
...,...,...,...,...,...,...,...,...,...
493531,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France,10.20
493532,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France,12.60
493533,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60
493534,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,16.60


In [37]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
Quantity,344771.0,10.50881,1.0,2.0,6.0,12.0,12540.0,38.140261
InvoiceDate,344771,2011-07-10 01:22:15.664310272,2010-12-01 08:34:00,2011-04-06 12:56:00,2011-07-29 09:48:00,2011-10-20 10:15:00,2011-12-09 12:50:00,NaN
UnitPrice,344771.0,2.862683,0.0,1.25,1.95,3.75,295.0,4.106272
CustomerID,344771.0,15335.498374,12347.0,13940.0,15361.0,16813.0,18287.0,1708.197516
Revenue,344771.0,17.444812,0.0,4.95,11.9,18.72,3861.0,36.483774


In [38]:
df.to_csv("../data/cleaned.csv", index=False)